File: 2011/QR/bot_trust/bt.clj
(require 'clojure.string)

(defn target_pos [color seqs org]
	(let [found (filter (fn [v] (= (first v) color)) seqs)]
		(if (empty? found)
			org
			(Integer/parseInt (last (first found))))))
	
(defn new_value [tv ov dist]
	(if (> tv ov) (+ ov dist) (- ov dist)))
		
(defn solve [seqs o b]
	(if (empty? seqs) 
		0
		(let [to (target_pos "O" seqs o) tb (target_pos "B" seqs b)
				job (first seqs)
				dist_o (Math/abs (- to o)) dist_b (Math/abs (- tb b))
				req_s (if (= (first job) "O") (+ 1 dist_o) (+ 1 dist_b))]		
			(+ req_s (solve (rest seqs) (new_value to o (min dist_o req_s)) (new_value tb b (min dist_b req_s)))))))		
	
(defn read-int-line [] (Integer/parseInt (read-line)))

(defn print-result [n value] 
	(println (str "Case #" (+ 1 n) ": " value)))
	
(defn makeq [str]
	(for [pair (re-seq #"\w \d+" str)] (seq (clojure.string/split pair #" "))))

(dotimes [no (read-int-line)] 
	(let [s (clojure.string/split (read-line) #" ")]
		(print-result no (solve (makeq (clojure.string/join " " (rest s))) 1 1))))

File: 2011/QR/bot_trust/bt.rb
def target_pos color, seqs
  v = seqs.assoc(color)
  v && v[1]
end

def solve seqs, o = 1, b = 1
  return 0 if seqs.empty?
  to = target_pos("O", seqs) || o
  tb = target_pos("B", seqs) || b
  job = seqs.shift
  dist_o = (to - o).abs
  dist_b = (tb - b).abs
  req_s = (job[0] == "O") ? dist_o + 1 : dist_b + 1
  dist_o = [dist_o, req_s].min
  dist_b = [dist_b, req_s].min
  no = (to > o) ? o + dist_o : o - dist_o
  nb = (tb > b) ? b + dist_b : b - dist_b
  req_s + solve(seqs, no, nb)
end

gets.to_i.times do |no|
  s = gets.split
  n = s.shift.to_i
  seqs = []
  n.to_i.times { seqs << [s.shift, s.shift.to_i] }
  puts "Case ##{no+1}: #{solve seqs}"  
end